File: 543. Diameter of Binary Tree/sol.py
class TreeNode:
    def __init__(self, val = 0 , left = Optional[TreeNode], right = Optional[TreeNode]):
        self.val = val
        self.left = left 
        self.right = right 
        


class Solution:
    def diameterOfBinaryTree(self, root: Optional[TreeNode]) -> int:
        
        
        def dfs(node: Optional[TreeNode]) -> (int, int):
            if not node:
                return 0, 0
            depth = 1
            left, maxLeft = dfs(node.left)
            right, maxRight = dfs(node.right)

            diameter = left + right
            print(left, maxLeft, right, maxRight)
            return 1 + max(left, right) , max(diameter, maxLeft, maxRight)

        depth, maxDepth = dfs(root)

        return maxDepth

File: 543. Diameter of Binary Tree/sol.go
package diameterOfBinaryTree

type TreeNode struct {
	Val int
	Left *TreeNode
	Right *TreeNode 

}
func diameterOfBinaryTree(root *TreeNode) int { 
    res := 0 
    
    var dfs func(node *TreeNode) (int) 
    dfs = func(node *TreeNode) (int) {
        if node == nil {
            return 0
        }
        
        leftHeight := dfs(node.Left)
        rightHeight := dfs(node.Right)

        maxDepth := max(leftHeight, rightHeight) + 1
        currDiameter := leftHeight + rightHeight
        res = max(res, currDiameter)

        return maxDepth
    }

    dfs(root)
    return res

}

func max(args ...int) int {
    m := args[0]
    for _, a := range args {
        if a > m {
            m = a
        }
    }
    return m
}